autogluon/utils/tabular/ml/models/abstract/model_trial.py

```diff
@@ -19,27 +20,33 @@ def model_trial(args, reporter):
                 'num_threads', 'num_gpus' to set specific resources in model.fit()
             - model.save() must have return_filename, file_prefix, directory options
     """
-    trial_id = args.pop('task_id')  # Note may not start at 0 if HPO has been run for other models with same scheduler
-    directory = args.pop('directory')  # TODO: Separate model parameters vs HPO
-    file_prefix = f"trial_{trial_id}_"  # append to all file names created during this trial. Do NOT change!
-    model = args.pop('model')  # the model object must be passed into model_trial() here
-
-    dataset_train_filename = args.pop('dataset_train_filename')
-    X_train, Y_train = pickle.load(open(directory + dataset_train_filename, 'rb'))
-    dataset_val_filename = args.pop('dataset_val_filename', None)
-    if dataset_val_filename is None:
-        raise NotImplementedError("must provide validation data for model_trial()")
-    X_val, Y_val = pickle.load(open(directory + dataset_val_filename, 'rb'))
-    model.params = model.params.copy()  # all hyperparameters must be stored in this dict
-    model.params.update(args)
-    if 'seed_value' in model.params:
-        random.seed(model.params['seed_value'])
-        np.random.seed(model.params['seed_value'])
-
-        model.params.pop('seed_value')  # TODO: Should we be popping this?
-
-    model.fit(X_train, Y_train, X_val, Y_val)
-    val_perf = model.score(X_val, Y_val)
-    trial_model_file = model.save(file_prefix=file_prefix, directory=directory, return_filename=True)
-    reporter(epoch=0, validation_performance=val_perf, directory=directory, file_prefix=file_prefix, 
-             trial_model_file=trial_model_file)  #  auxiliary_info = ?)
+    try:
```

Comment:
If 9 trials run successfully and then the next fails, will it keep training past the failure on new HPO configs or stop HPO entirely? If it stops HPO, are the 9 successfully trials returned or lost?

Reply:
Hi @Innixma , 

Good question. So we are lucky, because for each HPO call corresponds a separate Task and also job which is doing this task. Separate job means that it has separate reporters. So it is terminating only single HPO call reporter and the trials which are finished with success are not lost. You can check for example `autogluon.scheduler.fifo.FIFOScheduler.schedule_next`  and corresponding `autogluon.scheduler.fifo.FIFOScheduler.add_job`.

Please see modified code sample, where I put wrong name of `activation` function for NN options. 

```
import autogluon as ag
from autogluon.task import TabularPrediction as task

import sklearn.datasets
import pandas as pd

cancer_data = sklearn.datasets.load_breast_cancer(return_X_y=False)
data = pd.DataFrame(cancer_data['data'], columns=cancer_data['feature_names'])
data['target'] = cancer_data['target']

hp_tune = True
nn_options = {
    'num_epochs': 10,
    'activation': ag.space.Categorical('relu', 'hello_my_friend'),
}

hyperparameters = {
    'NN': nn_options
}

time_limits = 2 * 60
num_trials = 5

output_directory = './agModels--tuningCancerDataset'
data = task.Dataset(data)
label_column = 'target'

# case 1: random search
search_strategy = 'random'
print('\n' * 2)
print('#' * 100)
print(f'Case 1: Running HPO suing search strategy = "{search_strategy}".')
print()
predictor = task.fit(
    train_data=data,
    holdout_frac=0.3,
    label=label_column,
    output_directory=output_directory,
    time_limits=time_limits,
    num_trials=num_trials,
    hyperparameter_tune=hp_tune,
    hyperparameters=hyperparameters,
    search_strategy=search_strategy,
    auto_stack=False,
    stack_ensemble_levels=0
)

print('\n' * 2)
print('#' * 100)
print('HPO results:')
print(predictor.leaderboard(silent=True))
```

As a result I got:
![image](https://user-images.githubusercontent.com/28273805/75077252-7afe4080-5502-11ea-8bed-162a6b7e1e73.png)
So instead of 5 trials it has only 3, so it is not loosing history.

Reply:
Great, thanks for the explanation!

Reply:
Regarding cases where all models fail (bad config), will it keep training (and immediately failing) models until time_limit or max_trials is reached? We may want to add some cutoff in future if the first 5-10 trials all fail.

Reply:
It will try all of them and fail very fast. So it will not wait until time_limit or max_trials, but will try all n_trials. We can think about cut-off, but it really fails fast in case of wrong params.